drgn/cli.py

```diff
@@ -0,0 +1,151 @@
+    elif isinstance(value, (drgn.StackFrame, drgn.StackTrace, drgn.Type)):
+        text = str(value)
+    else:
+        text = repr(value)
+    try:
+        sys.stdout.write(text)
+    except UnicodeEncodeError:
+        encoded = text.encode(sys.stdout.encoding, "backslashreplace")
+        if hasattr(sys.stdout, "buffer"):
+            sys.stdout.buffer.write(encoded)
+        else:
+            text = encoded.decode(sys.stdout.encoding, "strict")
+            sys.stdout.write(text)
+    sys.stdout.write("\n")
+    setattr(builtins, "_", value)
+
+
+def run_cli(
+    prog: drgn.Program,
+    additional_globals: Optional[Dict[str, Any]] = None,
+    banner: Optional[str] = None,
+    additional_banner: Optional[str] = None,
+    quiet: bool = False,
+) -> None:
+    """
+    Run the drgn command line interface.
+
+    This function allows your application to embed a drgn CLI that operates the
+    same as the one which drgn itself provides. Simply provide an already
+    configured :class:`drgn.Program`, and any customizations you would like for
+    the banner and globals, and this function will create REPL and run it until
+    the user exits.
+
+    :param prog: The program to insert into the CLI namespace as ``prog``
+    :param additional_globals: A dictionary to add into the global namespace of
+      the CLI. Drgn will populate the namespace with the normal default: the
+      program, the drgn module, and relevant helpers. The additional globals
+      will be added last, and so they can override any preexisting variable with
+      the same name.
+    :param banner: Replace the normal drgn startup banner with your own. If you
+      simply want to add more information below drgn's banner, see
+      ``additional_banner`` below.
+    :param additional_banner: Add your own text which comes after drgn's default
+      banner
+    """
+    version = drgn_version_header()
+    sys.path.insert(0, "")
+
+    init_globals: Dict[str, Any] = {"prog": prog}
+    init_globals["drgn"] = drgn
+    drgn_globals = [
+        "NULL",
+        "Object",
+        "cast",
+        "container_of",
+        "execscript",
+        "offsetof",
+        "reinterpret",
+        "sizeof",
+    ]
+    for attr in drgn_globals:
+        init_globals[attr] = getattr(drgn, attr)
+    init_globals["__name__"] = "__main__"
+    init_globals["__doc__"] = None
+
+    histfile = os.path.expanduser("~/.drgn_history")
+    try:
+        readline.read_history_file(histfile)
+    except OSError as e:
+        if not isinstance(e, FileNotFoundError) and not quiet:
+            print("could not read history:", str(e), file=sys.stderr)
+
+    def write_history_file() -> None:
+        try:
+            readline.write_history_file(histfile)
+        except OSError as e:
+            if not quiet:
+                print("could not write history:", str(e), file=sys.stderr)
+
+    atexit.register(write_history_file)
```

Comment:
We should probably replace these `atexit` handlers with something that will be called before returning from this function rather than whenever the caller exits. `contextlib.ExitStack` might be useful.